capsules/extra/src/bus.rs

```diff
@@ -46,31 +56,163 @@ pub enum BusWidth {
     Bits64BE,
 }
 
-impl BusWidth {
+/// Each `BusAddr` struct represents a specific data width and endianness.
+pub struct BusAddr8(u8);
+pub struct BusAddr16BE(u16);
+pub struct BusAddr16LE(u16);
+pub struct BusAddr32BE(u32);
+pub struct BusAddr32LE(u32);
+pub struct BusAddr64BE(u64);
+pub struct BusAddr64LE(u64);
+
+impl From<BusAddr8> for BusAddr8080 {
+    fn from(value: BusAddr8) -> Self {
+        BusAddr8080::BusAddr8(value.0)
+    }
+}
+impl From<BusAddr16BE> for BusAddr8080 {
+    fn from(value: BusAddr16BE) -> Self {
+        BusAddr8080::BusAddr16BE(value.0)
+    }
+}
+impl From<BusAddr16LE> for BusAddr8080 {
+    fn from(value: BusAddr16LE) -> Self {
+        BusAddr8080::BusAddr16LE(value.0)
+    }
+}
+
+impl From<u8> for BusAddr8 {
+    fn from(value: u8) -> Self {
+        Self(value)
+    }
+}
+impl From<u16> for BusAddr16BE {
+    fn from(value: u16) -> Self {
+        Self(value)
+    }
+}
+impl From<u16> for BusAddr16LE {
+    fn from(value: u16) -> Self {
+        Self(value)
+    }
+}
+impl From<u32> for BusAddr32BE {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+impl From<u32> for BusAddr32LE {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+impl From<u64> for BusAddr64BE {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+impl From<u64> for BusAddr64LE {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+
+/// The `BusAddr` trait is implemented for each BusAddr struct.
+/// It provides information about the data width and a way
+/// to access the underlying byte representation.
+pub trait BusAddr {
+    const DATA_WIDTH: DataWidth;
+    fn len(&self) -> usize;
```

Comment:
What about having a default implementation here
```suggestion
    fn len(&self) -> usize {
      Self::DATA_WIDTH.width_in_bytes()
    }
```

or delete it completely and use `<addr as BusAddr>::DATA_WIDTH.width_in_bytes()`?